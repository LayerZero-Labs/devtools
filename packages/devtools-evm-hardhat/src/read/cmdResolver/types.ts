import type { BlockNumberTimeMarker, ResolvedTimestampTimeMarker, TimestampTimeMarker } from '@/read/types'

export interface ICommandResolverSdk {
    extractTimeMarkers(command: string): Promise<{
        blockNumberTimeMarkers: BlockNumberTimeMarker[]
        timestampTimeMarkers: TimestampTimeMarker[]
    }>

    resolveCmd(command: string, timeMarkers: ResolvedTimestampTimeMarker[]): Promise<string>
}

export * from './chain/types'
