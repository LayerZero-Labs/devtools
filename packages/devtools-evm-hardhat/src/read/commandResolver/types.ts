import type { BlockNumberTimeMarker, ResolvedTimestampTimeMarker, TimestampTimeMarker } from '@/read/types'
import type { Command } from '@layerzerolabs/lz-v2-utilities'

export interface ICommandResolverSdk {
    decodeCommand(command: string): Command

    extractTimeMarkers(command: string): Promise<{
        blockNumberTimeMarkers: BlockNumberTimeMarker[]
        timestampTimeMarkers: TimestampTimeMarker[]
    }>

    resolveCommand(command: string, timeMarkers: ResolvedTimestampTimeMarker[]): Promise<string>
}

export * from './chain/types'
