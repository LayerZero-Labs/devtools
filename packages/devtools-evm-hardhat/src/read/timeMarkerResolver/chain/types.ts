export interface ITimeMarkerResolverChainSdk {
    resolveTimestamps(timestamps: number[]): Promise<{ [timestamp: number]: number }>
}
