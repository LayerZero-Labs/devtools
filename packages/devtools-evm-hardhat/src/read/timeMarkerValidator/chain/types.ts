import type { ResolvedTimeMarker, ResolvedTimestampTimeMarker } from '@/read/types'

export interface ITimeMarkerValidatorChainSdk {
    checkResolvedTimeMarkerValidity(tms: Omit<ResolvedTimestampTimeMarker, 'eid'>[]): Promise<void> // throw if wrong timestamp

    assertTimeMarkerBlockConfirmations(tms: Omit<ResolvedTimeMarker, 'eid'>[]): Promise<void> // throw if not enough confirmations
}
