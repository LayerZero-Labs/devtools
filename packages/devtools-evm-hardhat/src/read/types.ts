export * from './common/types'
export * from './commandResolver/types'
export * from './timeMarkerResolver/types'
export * from './timeMarkerValidator/model'
