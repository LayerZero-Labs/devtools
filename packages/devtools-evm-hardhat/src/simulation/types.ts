import { AnvilOptions } from '@layerzerolabs/devtools-evm'

export interface SimulationUserConfig {
    /**
     * Simulation works by creating local forks of networks specified
     * in hardhat config. These networks are containerized and not accessible
     * from the developer machine - instead, a simple proxy server
     * listening on this port proxies requests to the individual networks.
     *
     * For example, a hardhat configuration with networks `fuji` and `mainnet`
     * will result in two forks being created. These networks will then be accessible
     * on `http://localhost:<port>/fuji` and `http://localhost:<port>/mainnet`
     *
     * @default 8545
     */
    port?: number

    /**
     * Simulation task stores its artifacts on the local filesystem.
     *
     * To customize the path, set this property to a relative or absolute path
     * (relative path will be resolved against the root path of your hardhat project)
     *
     * @default .layerzero
     */
    directory?: string

    /**
     * Simulation will by default overwrite the network accounts by values
     * provided in Anvil config (or defaults if not present).
     *
     * The mnemonic will default to "test test test test test test test test test test test junk"
     * and the number of funded accounts to 10.
     *
     * @default true
     */
    overwriteAccounts?: boolean

    /**
     * Anvil overrides for the underlying EVM nodes.
     */
    anvil?: SimulationAnvilUserConfig
}

/**
 * Resolved simulation config
 */
export interface SimulationConfig {
    port: number
    directory: string
    overwriteAccounts: boolean
    anvil: SimulationAnvilConfig
}

/**
 * Resolved simulation config for anvil.
 *
 * This config is created by taking the user config
 * and applying defaults.
 */
export type SimulationAnvilConfig = AnvilOptions & {
    mnemonic: NonNullable<AnvilOptions['mnemonic']>
}

/**
 * User facing simulation config for anvil.
 *
 * This config cannot override several system attributes for anvil
 * (since it would not really make sense seeing that anvil is being run in containers):
 *
 * - host & port: this would break the docker port bindings
 * - state: state is kept within the container and is not accessible from the outside
 * - forkUrl: this is set based on the hardhat config and points to the hardhat network url
 */
export type SimulationAnvilUserConfig = Omit<AnvilOptions, 'host' | 'port' | 'state' | 'forkUrl'>
