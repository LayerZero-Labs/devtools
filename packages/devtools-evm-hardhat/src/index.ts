// Hardhat type augmentation
import './type-extensions'

// Regular exports
export * from './artifacts'
export * from './config'
export * from './constants'
export * from './errors'
export * from './internal'
export * from './omnigraph'
export * from './provider'
export * from './runtime'
export * from './transactions'
export * from './types'
