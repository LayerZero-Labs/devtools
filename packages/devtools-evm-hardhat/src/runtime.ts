import type {
    HardhatRuntimeEnvironment,
    EthereumProvider,
    ConfigurableTaskDefinition,
    HardhatArguments,
} from 'hardhat/types'

import pMemoize from 'p-memoize'
import type { JsonRpcProvider } from '@ethersproject/providers'
import { ConfigurationError } from './errors'
import { HardhatContext } from 'hardhat/internal/context'
import { loadConfigAndTasks } from 'hardhat/internal/core/config/config-loading'
import { Environment as HardhatRuntimeEnvironmentImplementation } from 'hardhat/internal/core/runtime-environment'
import { EndpointId } from '@layerzerolabs/lz-definitions'
import { EndpointBasedFactory, Factory, formatEid } from '@layerzerolabs/devtools'
import { EthersProviderWrapper } from '@nomiclabs/hardhat-ethers/internal/ethers-provider-wrapper'
import assert from 'assert'
import memoize from 'micro-memoize'
import { subtask, task } from 'hardhat/config'

/**
 * Helper type for when we need to grab something asynchronously by the network name
 */
export type GetByNetwork<TValue> = Factory<[networkName: string], TValue>

/**
 * Helper type for when we need to grab something asynchronously by the network name
 */
export type GetByEid<TValue> = Factory<[eid: EndpointId], TValue>

/**
 * Creates and sets the default hardhat context.
 *
 * This function will fail if a context has already been created
 */
export const createDefaultContext = (hardhatArguments: Partial<HardhatArguments> = {}) => {
    const ctx = HardhatContext.createHardhatContext()
    const resolvedArguments: HardhatArguments = {
        showStackTraces: false,
        version: false,
        help: false,
        emoji: false,
        verbose: false,
        ...hardhatArguments,
    }
    const { resolvedConfig, userConfig } = loadConfigAndTasks(resolvedArguments)
    const envExtenders = ctx.environmentExtenders
    const providerExtenders = ctx.providerExtenders
    const taskDefinitions = ctx.tasksDSL.getTaskDefinitions()
    const scopesDefinitions = ctx.tasksDSL.getScopesDefinitions()
    const env = new HardhatRuntimeEnvironmentImplementation(
        resolvedConfig,
        resolvedArguments,
        taskDefinitions,
        scopesDefinitions,
        envExtenders,
        ctx.experimentalHardhatNetworkMessageTraceHooks,
        userConfig,
        providerExtenders
    )

    ctx.setHardhatRuntimeEnvironment(env as unknown as HardhatRuntimeEnvironment)
}

/**
 * Returns the default hardhat context for the project, i.e.
 * the context that the project has been setup with.
 *
 * Throws if there is no context.
 *
 * @returns {HardhatContext}
 */
export const getDefaultContext = (): HardhatContext => {
    // Context is registered globally as a singleton and can be accessed
    // using the static methods of the HardhatContext class
    //
    // In our case we require the context to exist, the other option would be
    // to create it and set it up - see packages/hardhat-core/src/register.ts for an example setup
    try {
        return HardhatContext.getHardhatContext()
    } catch (error: unknown) {
        throw new ConfigurationError(`Could not get Hardhat context: ${error}`)
    }
}

/**
 * Returns the default `HardhatRuntimeEnvironment` (`hre`) for the project.
 *
 * Throws if there is no `HardhatRuntimeEnvironment`.
 *
 * @returns {HardhatRuntimeEnvironment}
 */
export const getDefaultRuntimeEnvironment = (): HardhatRuntimeEnvironment => {
    // The first step is to get the hardhat context
    const context = getDefaultContext()

    // We require the hardhat environment to already exist
    //
    // Again, we could create it but that means we'd need to duplicate the bootstrap code
    // that hardhat does when setting up the environment
    try {
        return context.getHardhatRuntimeEnvironment()
    } catch (error: unknown) {
        throw new ConfigurationError(`Could not get Hardhat Runtime Environment: ${error}`)
    }
}

/**
 * Creates a clone of the HardhatRuntimeEnvironment for a particular network
 *
 * ```typescript
 * const env = getHreByNetworkName("bsc-testnet");
 *
 * // All the ususal properties are present
 * env.deployments.get("MyContract")
 * ```
 *
 * @returns {Promise<HardhatRuntimeEnvironment>}
 */
export const getHreByNetworkName: GetByNetwork<HardhatRuntimeEnvironment> = pMemoize(async (networkName) => {
    const context = getDefaultContext()
    const environment = getDefaultRuntimeEnvironment()

    try {
        // The last step is to create a duplicate enviornment that mimics the original one
        // with one crucial difference - the network setup
        return new HardhatRuntimeEnvironmentImplementation(
            environment.config,
            {
                ...environment.hardhatArguments,
                network: networkName,
            },
            environment.tasks,
            environment.scopes,
            context.environmentExtenders,
            context.experimentalHardhatNetworkMessageTraceHooks,
            environment.userConfig,
            context.providerExtenders
            // This is a bit annoying - the environmentExtenders are not stronly typed
            // so TypeScript complains that the properties required by HardhatRuntimeEnvironment
            // are not present on HardhatRuntimeEnvironmentImplementation
        ) as unknown as HardhatRuntimeEnvironment
    } catch (error: unknown) {
        throw new ConfigurationError(`Could not setup Hardhat Runtime Environment: ${error}`)
    }
})

/**
 * Creates a clone of the HardhatRuntimeEnvironment for a particular network
 * identified by endpoint ID
 *
 * ```typescript
 * const getHreByEid = createGetHreByEid()
 * const env = await getHreByEid(EndpointId.AVALANCHE_V2_TESTNET);
 *
 * // All the ususal properties are present
 * env.deployments.get("MyContract")
 * ```
 *
 * @returns {Promise<HardhatRuntimeEnvironment>}
 */
export const createGetHreByEid = (
    hre = getDefaultRuntimeEnvironment()
): EndpointBasedFactory<HardhatRuntimeEnvironment> =>
    pMemoize(async (eid: EndpointId) => getHreByNetworkName(getNetworkNameForEid(eid, hre)))

/**
 * Helper function that wraps an EthereumProvider with EthersProviderWrapper
 * so that we can use it further with ethers as a regular JsonRpcProvider
 *
 * @param {EIP1193Provider} provider
 * @returns {JsonRpcProvider}
 */
export const wrapEIP1193Provider = (provider: EthereumProvider): JsonRpcProvider => new EthersProviderWrapper(provider)

/**
 * Gets an EndpointId defined in the hardhat config
 * for a particular network name (as an `eid` property).
 *
 * Throws if the network or the eid are not defined
 *
 * @param {string} networkName
 * @param {HardhatRuntimeEnvironment | undefined} [hre]
 * @returns {EndpointId}
 */
export const getEidForNetworkName = (
    networkName: string,
    hre: HardhatRuntimeEnvironment = getDefaultRuntimeEnvironment()
): EndpointId => {
    const networkConfig = hre.config.networks[networkName]
    assert(networkConfig, `Network '${networkName}' is not defined in hardhat config`)
    assert(networkConfig.eid != null, `Network '${networkName}' does not have 'eid' property defined in its config`)

    return networkConfig.eid
}

/**
 * Gets a network name with its `eid` property matching
 * a particular `eid`
 *
 * Throws if there is no such network or if there are multiple
 * networks defined with the same `eid`
 *
 * @param {EndpointId} eid
 * @param {HardhatRuntimeEnvironment | undefined} [hre]
 * @returns {string}
 */
export const getNetworkNameForEid = (
    eid: EndpointId,
    hre: HardhatRuntimeEnvironment = getDefaultRuntimeEnvironment()
): string => {
    // We are using getEidsByNetworkName to get the nice validation of network config
    const eidsByNetworkName = getEidsByNetworkName(hre)

    for (const [networkName, networkEid] of Object.entries(eidsByNetworkName)) {
        if (networkEid === eid) {
            return networkName
        }
    }

    // Here we error out if there are no networks with this eid
    assert(false, `Could not find a network for eid ${eid} (${formatEid(eid)})`)
}

/**
 * Gets a record containing the mapping between network names and endpoint IDs.
 * Will also return the network names for which the `eid` has not been defined
 *
 * Throws if there are multiple networks defined with the same `eid`
 *
 * @param {HardhatRuntimeEnvironment | undefined} [hre]
 * @returns {Record<string, EndpointId | undefined>}
 */
export const getEidsByNetworkName = memoize(
    (hre: HardhatRuntimeEnvironment = getDefaultRuntimeEnvironment()): Record<string, EndpointId | undefined> => {
        // First we get the network name -> network config pairs
        const networkEntries = Object.entries(hre.config.networks)
        // And map the network config to an endpoint ID
        const eidEntries = networkEntries.map(
            ([networkName, networkConfig]) => [networkName, networkConfig.eid] as const
        )
        // Now we turn the entries back into a record
        const eidsByNetworkName = Object.fromEntries(eidEntries)

        // Now we check that the user has not configured the endpoint ID mapping incorrectly
        // (i.e. there are more networks configured with the same endpoint ID)
        //
        // For this we'll drop all the networks whose endpoint IDs are not defined
        const eidEntriesWithDefinedEid = eidEntries.filter(([_, eid]) => eid != null)
        const definedEidsByNetworkName = Object.fromEntries(eidEntriesWithDefinedEid)

        // Now we grab the sets of unique network names and endpoint IDs
        const allDefinedEids = new Set(Object.values(definedEidsByNetworkName))
        const allNetworkNames = new Set(Object.keys(definedEidsByNetworkName))

        // If the number of unique networks matches the number of unique endpoint IDs, there are no duplicates
        if (allDefinedEids.size === allNetworkNames.size) {
            return eidsByNetworkName
        }

        // At this point the number of defined endpoint IDs can only be lower than
        // the number of defined network names (since network names are taken from the keys
        // of an object and endpoint IDs from its values)
        //
        // To let the user know whihc networks to fix, we need to grab all the ones that
        // have been duplicated
        //
        // We are not claiming any efficiency of this algorithm as we don't expect any large numbers of networks
        const duplicatedNetworkNames = Array.from(allDefinedEids)
            // First we grab all the network names with this endpoint ID
            .map((eid) =>
                eidEntriesWithDefinedEid.flatMap(([networkName, definedEid]) =>
                    eid === definedEid ? [networkName] : []
                )
            )
            // Then we find all the network names listed more than once
            .filter((networkNames) => networkNames.length > 1)

        // Now we let the user know which network names have identical endpoint IDs
        const messages = duplicatedNetworkNames
            .map(
                (networkNames) =>
                    `- ${networkNames.join(', ')} have eid set to ${formatEid(eidsByNetworkName[networkNames[0]!]!)}`
            )
            .join('\n')

        throw new Error(
            `Found multiple networks configured with the same 'eid':\n\n${messages}\n\nPlease fix this in your hardhat config.`
        )
    }
)

/**
 * Helper utility that copies the whole task definition under a new name
 *
 * This is useful if a new task needs to have the same interface as an existing task,
 * for example if we want to create a slightly modified version of a wire task
 * without needing to retype all the `.addFlag` and `.addOption`
 *
 * @param {string} parentTaskName Task to inherit the options and the action from
 * @param {HardhatRuntimeEnvironment} [hre]
 * @returns {(taskName: string) => ConfigurableTaskDefinition}
 */
export const inheritTask =
    (parentTaskName: string, context = getDefaultContext()) =>
    (taskName: string): ConfigurableTaskDefinition => {
        // For now we only support non-scoped tasks
        const parentTaskDefinition = context.tasksDSL.getTaskDefinition(undefined, parentTaskName)
        assert(parentTaskDefinition != null, `Missing task definition for ${parentTaskName}`)

        // First we create the task definition itself
        const creator = parentTaskDefinition.isSubtask ? subtask : task
        const childTask = creator(taskName).setAction(parentTaskDefinition.action)

        // Then we start setting properties
        if (parentTaskDefinition.description != null) {
            childTask.setDescription(parentTaskDefinition.description)
        }

        // Params go first (just because I said so, not for any particular reason)
        for (const definition of Object.values(parentTaskDefinition.paramDefinitions)) {
            // Params need to be treated based on their type (flag/param)
            if (definition.isFlag) {
                childTask.addFlag(definition.name, definition.description)
            } else {
                childTask.addParam(
                    definition.name,
                    definition.description,
                    definition.defaultValue,
                    definition.type,
                    definition.isOptional
                )
            }
        }

        // Positional params go second
        for (const definition of parentTaskDefinition.positionalParamDefinitions) {
            if (definition.isVariadic) {
                childTask.addVariadicPositionalParam(
                    definition.name,
                    definition.description,
                    definition.defaultValue,
                    definition.type,
                    definition.isOptional
                )
            } else {
                childTask.addPositionalParam(
                    definition.name,
                    definition.description,
                    definition.defaultValue,
                    definition.type,
                    definition.isOptional
                )
            }
        }

        return childTask
    }
