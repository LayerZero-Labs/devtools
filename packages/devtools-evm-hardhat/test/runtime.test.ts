import hre from 'hardhat'
import { DeploymentsManager } from 'hardhat-deploy/dist/src/DeploymentsManager'
import { createGetHreByEid, getEidForNetworkName, getNetworkNameForEid, getHreByNetworkName } from '@/runtime'
import type { DeploymentSubmission } from 'hardhat-deploy/dist/types'
import { EndpointId } from '@layerzerolabs/lz-definitions'
import { HardhatRuntimeEnvironment } from 'hardhat/types'

jest.spyOn(DeploymentsManager.prototype, 'getChainId').mockResolvedValue('1')

describe('runtime', () => {
    describe('getHreByNetworkName', () => {
        it('should reject with an invalid network', async () => {
            await expect(getHreByNetworkName('not-in-hardhat-config')).rejects.toBeTruthy()
        })

        it('should return a HardhatRuntimeEnvironment with correct network', async () => {
            const runtime = await getHreByNetworkName('ethereum-mainnet')

            expect(runtime.network.name).toEqual('ethereum-mainnet')
            expect(runtime.deployments).toMatchObject({
                get: expect.any(Function),
            })
        })

        it('should have the config setup correctly', async () => {
            const ethRuntime = await getHreByNetworkName('ethereum-mainnet')
            const bscRuntime = await getHreByNetworkName('bsc-testnet')

            expect(ethRuntime.network.name).toBe('ethereum-mainnet')
            expect(ethRuntime.network.config).toBe(hre.config.networks['ethereum-mainnet'])
            expect(bscRuntime.network.name).toBe('bsc-testnet')
            expect(bscRuntime.network.config).toEqual(hre.config.networks['bsc-testnet'])
        })

        it('should save the deployment to correct network', async () => {
            const bscRuntime = await getHreByNetworkName('bsc-testnet')
            const ethRuntime = await getHreByNetworkName('ethereum-mainnet')
            const now = Date.now()
            const deploymentSubmission = {
                args: ['bsc-testnet', now],
            } as DeploymentSubmission

            // First we want to save the deployment for bsc-testnet
            await bscRuntime.deployments.save('Mock', deploymentSubmission)

            // Then we check whether it has been saved
            const deployment = await bscRuntime.deployments.get('Mock')
            expect(deployment.args).toEqual(deploymentSubmission.args)

            // And finally we check whether it was not by accident saved for ethereum-mainnet
            const nonExistentDeployment = await ethRuntime.deployments.getOrNull('Mock')
            expect(nonExistentDeployment?.args).not.toEqual(deploymentSubmission.args)
        })
    })

    describe('createGetHreByEid()', () => {
        it('should reject with an endpoint that is not in the hardhat config', async () => {
            await expect(createGetHreByEid(hre)(EndpointId.CATHAY_TESTNET)).rejects.toThrow(
                'Could not find a network for eid 10171 (CATHAY_TESTNET)'
            )
        })

        it('should return a HardhatRuntimeEnvironment with correct network', async () => {
            const runtime = await createGetHreByEid(hre)(EndpointId.ETHEREUM_V2_MAINNET)

            expect(runtime.network.name).toEqual('ethereum-mainnet')
            expect(runtime.deployments).toMatchObject({
                get: expect.any(Function),
            })
        })
    })

    describe('getEidForNetworkName', () => {
        it('should throw if there is no such network defined', () => {
            expect(() => getEidForNetworkName('nonsense')).toThrow(
                `Network 'nonsense' is not defined in hardhat config`
            )
        })

        it('should throw if the network does not have eid defined', () => {
            expect(() => getEidForNetworkName('bsc-testnet')).toThrow(
                `Network 'bsc-testnet' does not have 'eid' property defined in its config`
            )
        })

        it('should return eid if defined', () => {
            expect(getEidForNetworkName('ethereum-testnet')).toBe(EndpointId.ETHEREUM_TESTNET)
            expect(getEidForNetworkName('ethereum-mainnet')).toBe(EndpointId.ETHEREUM_V2_MAINNET)
        })
    })

    describe('getNetworkNameForEid', () => {
        it('should throw if there is no such eid defined', () => {
            expect(() => getNetworkNameForEid(EndpointId.CATHAY_TESTNET)).toThrow(
                'Could not find a network for eid 10171 (CATHAY_TESTNET)'
            )
        })

        it('should throw if there are more networks defined with the same eid', () => {
            const hre = {
                config: {
                    networks: {
                        one: {
                            eid: EndpointId.APTOS_MAINNET,
                        },
                        two: {
                            eid: EndpointId.APTOS_MAINNET,
                        },
                        three: {
                            eid: EndpointId.APTOS_MAINNET,
                        },
                    },
                },
            }

            expect(() =>
                getNetworkNameForEid(EndpointId.APTOS_MAINNET, hre as unknown as HardhatRuntimeEnvironment)
            ).toThrow(`Multiple networks found with 'eid' set to 108 (APTOS_MAINNET): one, two, three`)
        })
    })
})
