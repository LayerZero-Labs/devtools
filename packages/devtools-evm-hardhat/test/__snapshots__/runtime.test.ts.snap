// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runtime getEidsByNetworkName should throw if there are more networks defined with the same eid 1`] = `
"Found multiple networks configured with the same 'eid':

- one, two, three have eid set to APTOS_MAINNET
- four, five have eid set to BASE_TESTNET

Please fix this in your hardhat config."
`;

exports[`runtime getNetworkNameForEid should throw if there are more networks defined with the same eid 1`] = `
"Found multiple networks configured with the same 'eid':

- one, two, three have eid set to APTOS_MAINNET

Please fix this in your hardhat config."
`;
